File: builtin/runtime/mcp.lua
-- builtin/runtime/mcp.lua --- T M9.1 friendly Lua surface for pmacs.mcp.*
--
-- The Rust binding (`pmacs.mcp._send_request_raw`) returns a raw
-- async-runtime JobId. Package code wants the same dispatch shape
-- as `pmacs.workers.compute_sum`, `pmacs.fs.read_dir`, and friends:
--
--   pmacs.async(function()
--     local result = pmacs.mcp.send_request(server, "ping", {}):await()
--     ...
--   end)
--
-- This file replaces `pmacs.mcp.send_request` with the Handle-
-- returning wrapper. The Rust manager registers each request with
-- the async runtime and settles the job when the JSON-RPC response
-- lands on the supervisor pipe; the Handle's :await() resumes the
-- parked coroutine with the response's `result` value (already
-- translated to a Lua table) — or raises `{ tag = "failed",
-- message = ... }` if the server returned a JSON-RPC error, or
-- `{ tag = "cancelled", id = ... }` if the awaiter cancelled or the
-- server died before responding.

local mcp_mod = pmacs.mcp
assert(mcp_mod, "pmacs.mcp must be installed before mcp.lua loads")
assert(mcp_mod._send_request_raw,
  "pmacs.mcp._send_request_raw missing; lua_bindings::install_mcp not run?")

local workers_mod = pmacs.workers
assert(workers_mod and workers_mod._new_handle,
  "pmacs.workers._new_handle missing; did async.lua load before mcp.lua?")

local raw_send_request = mcp_mod._send_request_raw
local new_handle = workers_mod._new_handle

-- pmacs.mcp.send_request(server, method, params) -> Handle
--
-- `server` is the McpServerIdLua handle returned by pmacs.mcp.spawn.
-- `method` is a string. `params` is an optional table (or nil). The
-- return value is a Handle that completes with the response's
-- `result` table.
function mcp_mod.send_request(server, method, params)
  if type(method) ~= "string" then
    error("pmacs.mcp.send_request: method must be a string, got " .. type(method))
  end
  if params ~= nil and type(params) ~= "table" then
    error("pmacs.mcp.send_request: params must be a table or nil, got " .. type(params))
  end
  local job_id = raw_send_request(server, method, params)
  return new_handle(job_id)
end

-- T M9.2: pmacs.mcp.read_resource(server, uri) -> Handle
--
-- Cache-aware MCP resource fetch. Three observable outcomes, all
-- delivered through the same Handle shape:
--
--   * cache hit: handle settles with the cached result (one tick late)
--   * in-flight coalesce: handle attaches to an existing in-flight
--     request, settles with the same result
--   * cache miss: dispatches a fresh `resources/read`, settles with
--     the response
--
-- `pmacs.mcp.invalidate_resource(server, uri)` (installed directly
-- by Rust as a non-Handle function) drops the cache entry; subsequent
-- read_resource calls re-dispatch.
local raw_read_resource = mcp_mod._read_resource_raw
assert(raw_read_resource,
  "pmacs.mcp._read_resource_raw missing; lua_bindings::install_mcp not run?")

function mcp_mod.read_resource(server, uri)
  if type(uri) ~= "string" then
    error("pmacs.mcp.read_resource: uri must be a string, got " .. type(uri))
  end
  local job_id = raw_read_resource(server, uri)
  return new_handle(job_id)
end

-- T M9.3: pmacs.mcp.invoke_tool(server, name, args) -> Handle
--
-- Tool invocation via JSON-RPC `tools/call`. The handle settles
-- with the response's `result` table (`{ content = [...],
-- isError = false }` shape from the MCP spec) on success, or
-- raises a Lua error on either of two failure paths:
--
--   * JSON-RPC error response from the server: standard async
--     failure path.
--   * MCP "tool errored" success response (`isError: true`):
--     translated by the manager into a Failed outcome with the
--     extracted text content as the message. This is a deliberate
--     API choice (see M9.3 audit) — invoke_tool's contract is
--     "raises Lua errors on tool failure", so callers don't have
--     to write `if r.isError then ... end` boilerplate at every
--     call site.
--
-- Callers needing structured access to the raw `{isError, content}`
-- table use `pmacs.mcp.send_request(server, "tools/call", { name = ...,
-- arguments = ... })` to bypass the translator.
local raw_invoke_tool = mcp_mod._invoke_tool_raw
assert(raw_invoke_tool,
  "pmacs.mcp._invoke_tool_raw missing; lua_bindings::install_mcp not run?")

function mcp_mod.invoke_tool(server, name, args)
  if type(name) ~= "string" then
    error("pmacs.mcp.invoke_tool: name must be a string, got " .. type(name))
  end
  if args ~= nil and type(args) ~= "table" then
    error("pmacs.mcp.invoke_tool: args must be a table or nil, got " .. type(args))
  end
  local job_id = raw_invoke_tool(server, name, args)
  return new_handle(job_id)
end

-- T M9.4: pmacs.mcp.get_prompt(server, name [, args]) -> Handle
--
-- Resolve a prompt template via JSON-RPC `prompts/get`. The handle
-- settles with the response's `result` table:
--
--   { description = "...", messages = [{ role = "user", content = ... }, ...] }
--
-- on success, or raises a Lua error (via the async runtime's
-- `tag = "failed"` path) on JSON-RPC errors — including the
-- "missing required argument" case (`-32602`) which is how MCP
-- servers report missing args.
--
-- Unlike `invoke_tool`, there is no `isError`-style translator —
-- `prompts/get` has no semantic-failure path in the MCP spec; either
-- the prompt resolves (success) or the protocol-level call fails
-- (Lua error).
--
-- `args` may be omitted, nil, or an empty table — all three send
-- `arguments: {}` on the wire (the MCP spec requires the field).
local raw_get_prompt = mcp_mod._get_prompt_raw
assert(raw_get_prompt,
  "pmacs.mcp._get_prompt_raw missing; lua_bindings::install_mcp not run?")

function mcp_mod.get_prompt(server, name, args)
  if type(name) ~= "string" then
    error("pmacs.mcp.get_prompt: name must be a string, got " .. type(name))
  end
  if args ~= nil and type(args) ~= "table" then
    error("pmacs.mcp.get_prompt: args must be a table or nil, got " .. type(args))
  end
  local job_id = raw_get_prompt(server, name, args)
  return new_handle(job_id)
end

-- T M9.5: notification dispatcher.
--
-- pmacs.mcp.on_notification(method, fn) -> token
--
-- Register `fn` to be called whenever an MCP server emits a
-- notification with the given method. Multiple handlers per method
-- are allowed; they fire in registration order. The handler
-- receives `(server, params)` where `server` is an McpServerIdLua
-- and `params` is the notification's params table.
--
-- Returns an opaque token; pass it to `pmacs.mcp.off_notification`
-- to unregister.
--
-- The dispatcher hooks into pmacs._async.tick: on each tick, the
-- Rust manager's drain_notifications API is called, and pending
-- notifications are dispatched to registered handlers. Single
-- per-tick walk regardless of how many packages have registered;
-- M9.5/M9.6/M9.7 all share the mechanism.
local subscribe_raw = mcp_mod._subscribe_notification
local unsubscribe_raw = mcp_mod._unsubscribe_notification
local drain_raw = mcp_mod._drain_notifications
assert(subscribe_raw and unsubscribe_raw and drain_raw,
  "pmacs.mcp._subscribe_notification / _unsubscribe_notification / _drain_notifications missing")

-- Per-method handler list. `_handlers[method]` is a table where
-- entries are `{ token = <unique>, fn = <function> }`. Tokens are
-- monotonic so unregistration is unambiguous.
local _handlers = {}
local _next_token = 1

function mcp_mod.on_notification(method, fn)
  if type(method) ~= "string" then
    error("pmacs.mcp.on_notification: method must be a string, got " .. type(method))
  end
  if type(fn) ~= "function" then
    error("pmacs.mcp.on_notification: fn must be a function, got " .. type(fn))
  end
  local list = _handlers[method]
  if list == nil then
    list = {}
    _handlers[method] = list
    -- First handler for this method — register interest with the
    -- Rust manager so notifications/<method> get queued.
    subscribe_raw(method)
  end
  local token = _next_token
  _next_token = _next_token + 1
  list[#list + 1] = { token = token, fn = fn }
  return token
end

function mcp_mod.off_notification(method, token)
  local list = _handlers[method]
  if list == nil then return end
  for i, entry in ipairs(list) do
    if entry.token == token then
      table.remove(list, i)
      break
    end
  end
  if #list == 0 then
    _handlers[method] = nil
    -- Last handler dropped — tell Rust to stop queuing.
    unsubscribe_raw(method)
  end
end

-- Tick hook: drain queued notifications and dispatch.
local _orig_async_tick = pmacs._async.tick
function pmacs._async.tick()
  _orig_async_tick()
  local drained = drain_raw()
  for method, entries in pairs(drained) do
    local list = _handlers[method]
    if list ~= nil then
      for _, entry in ipairs(entries) do
        for _, handler in ipairs(list) do
          local ok, err = pcall(handler.fn, entry.server, entry.params)
          if not ok and pmacs.error then
            pmacs.error("pmacs.mcp.on_notification(" .. method ..
              ") handler raised: " .. tostring(err))
          end
        end
      end
    end
  end
end

File: src/lua_bindings/mcp.rs
// lua_bindings/mcp.rs --- pmacs.mcp: MCP client surface (T M9.1).

//! `pmacs.mcp.*` — Model Context Protocol client bindings. Split out of
//! `lua_bindings.rs` verbatim (audit F-016); behavior unchanged. Reaches
//! the shared JSON converters (still physically in the `lsp` section of
//! `mod.rs`) via `super::`.

use std::cell::RefCell;
use std::rc::Rc;

use mlua::{FromLua, Lua, Table, UserData, UserDataMethods, Value};

use super::{SharedProcessSupervisor, json_to_lua, lua_to_json};

use crate::mcp::{
    McpClientState, McpError, McpEvent, McpEventKind, McpManager, McpRestartPolicy, McpServerId,
    McpServerSpec, SharedMcpManager, state_label_for as mcp_state_label_for,
};

/// Lua-facing wrapper around [`McpServerId`]. Mirrors
/// [`LspServerIdLua`].
#[derive(Copy, Clone)]
pub struct McpServerIdLua(pub McpServerId);

impl McpServerIdLua {
    /// The wrapped [`McpServerId`].
    #[must_use]
    pub fn id(self) -> McpServerId {
        self.0
    }
}

impl FromLua for McpServerIdLua {
    fn from_lua(value: Value, _: &Lua) -> mlua::Result<Self> {
        match value {
            Value::UserData(ud) => Ok(*ud.borrow::<Self>()?),
            other => Err(mlua::Error::FromLuaConversionError {
                from: other.type_name(),
                to: "McpServerIdLua".to_string(),
                message: Some("expected an MCP server handle".to_string()),
            }),
        }
    }
}

impl UserData for McpServerIdLua {
    fn add_methods<M: UserDataMethods<Self>>(methods: &mut M) {
        methods.add_meta_method(mlua::MetaMethod::ToString, |_, this, ()| {
            Ok(format!("{}", this.0))
        });
        methods.add_meta_method(mlua::MetaMethod::Eq, |_, this, other: McpServerIdLua| {
            Ok(this.0 == other.0)
        });
        methods.add_method("raw", |_, this, ()| Ok(this.0.raw()));
    }
}

fn parse_mcp_restart(name: &str) -> mlua::Result<McpRestartPolicy> {
    Ok(match name {
        "never" | "Never" => McpRestartPolicy::Never,
        "on_crash" | "OnCrash" | "on-crash" => McpRestartPolicy::OnCrash,
        "always" | "Always" => McpRestartPolicy::Always,
        other => {
            return Err(mlua::Error::external(format!(
                "unknown MCP restart policy: {other:?} (expected never|on_crash|always)"
            )));
        }
    })
}

fn lua_to_mcp_spec(t: &Table) -> mlua::Result<McpServerSpec> {
    let label: String = t.get("label").unwrap_or_else(|_| "unnamed".to_owned());
    let command: String = t.get("command")?;
    let args: Vec<String> = t.get("args").unwrap_or_default();
    let cwd: Option<String> = t.get("cwd").ok().flatten();
    let env_t: Option<Table> = t.get("env").ok().flatten();
    let mut env: Vec<(String, String)> = Vec::new();
    if let Some(env_t) = env_t {
        env_t.for_each(|k: String, v: String| {
            env.push((k, v));
            Ok(())
        })?;
    }
    let restart = match t.get::<Option<String>>("restart").ok().flatten() {
        Some(s) => parse_mcp_restart(&s)?,
        None => McpRestartPolicy::OnCrash,
    };
    Ok(McpServerSpec {
        label,
        command,
        args,
        cwd: cwd.map(std::path::PathBuf::from),
        env,
        restart,
    })
}

fn mcp_state_to_lua(lua: &Lua, state: &McpClientState) -> mlua::Result<Table> {
    let t = lua.create_table_with_capacity(0, 5)?;
    t.set("kind", mcp_state_label_for(state))?;
    match state {
        McpClientState::Starting
        | McpClientState::Stopped { .. }
        | McpClientState::ShuttingDown => {}
        McpClientState::Initializing {
            init_request_id, ..
        } => {
            t.set("init_request_id", *init_request_id)?;
        }
        McpClientState::Initialized {
            capabilities,
            server_info,
            protocol_version,
            ..
        } => {
            t.set("capabilities", json_to_lua(lua, capabilities)?)?;
            if let Some(info) = server_info {
                t.set("server_info", json_to_lua(lua, info)?)?;
            }
            if let Some(v) = protocol_version {
                t.set("protocol_version", v.as_str())?;
            }
        }
        McpClientState::Crashed { reason, .. } => {
            t.set("reason", reason.as_str())?;
        }
    }
    Ok(t)
}

fn mcp_event_to_lua(lua: &Lua, ev: &McpEvent) -> mlua::Result<Table> {
    let t = lua.create_table_with_capacity(0, 5)?;
    t.set("server", McpServerIdLua(ev.server))?;
    match &ev.kind {
        McpEventKind::Started { pid } => {
            t.set("kind", "started")?;
            t.set("pid", *pid)?;
        }
        McpEventKind::Initialized { capabilities } => {
            t.set("kind", "initialized")?;
            t.set("capabilities", json_to_lua(lua, capabilities)?)?;
        }
        McpEventKind::Notification { method, params } => {
            t.set("kind", "notification")?;
            t.set("method", method.as_str())?;
            t.set("params", json_to_lua(lua, params)?)?;
        }
        McpEventKind::Request { id, method, params } => {
            t.set("kind", "request")?;
            t.set("request_id", json_to_lua(lua, id)?)?;
            t.set("method", method.as_str())?;
            t.set("params", json_to_lua(lua, params)?)?;
        }
        McpEventKind::Response {
            id,
            result,
            error,
            method,
        } => {
            t.set("kind", "response")?;
            t.set("request_id", *id)?;
            t.set("method", method.as_str())?;
            t.set("result", json_to_lua(lua, result)?)?;
            if let Some(e) = error {
                let err_t = lua.create_table_with_capacity(0, 3)?;
                err_t.set("code", e.code)?;
                err_t.set("message", e.message.as_str())?;
                if let Some(d) = &e.data {
                    err_t.set("data", json_to_lua(lua, d)?)?;
                }
                t.set("error", err_t)?;
            }
        }
        McpEventKind::ShuttingDown => {
            t.set("kind", "shutting_down")?;
        }
        McpEventKind::Stopped => {
            t.set("kind", "stopped")?;
        }
        McpEventKind::Crashed { reason } => {
            t.set("kind", "crashed")?;
            t.set("reason", reason.as_str())?;
        }
        McpEventKind::Restarting { attempt } => {
            t.set("kind", "restarting")?;
            t.set("attempt", *attempt)?;
        }
        McpEventKind::Stderr(bytes) => {
            t.set("kind", "stderr")?;
            t.set("bytes", lua.create_string(bytes)?)?;
        }
        McpEventKind::ProtocolError { message } => {
            t.set("kind", "protocol_error")?;
            t.set("message", message.as_str())?;
        }
    }
    Ok(t)
}

/// Lua → [`McpError`]. Either a plain string (becomes a generic
/// `code = -32603` internal error) or a table with `{code, message,
/// data?}`.
fn lua_to_mcp_error(value: Value) -> mlua::Result<McpError> {
    match value {
        Value::String(s) => Ok(McpError {
            code: -32603,
            message: s.to_str()?.to_owned(),
            data: None,
        }),
        Value::Table(t) => {
            let code: i64 = t.get("code").unwrap_or(-32603);
            let message: String = t.get("message").unwrap_or_else(|_| "error".to_owned());
            let data = match t.get::<Option<Value>>("data")? {
                Some(Value::Nil) | None => None,
                Some(other) => Some(lua_to_json(other)?),
            };
            Ok(McpError {
                code,
                message,
                data,
            })
        }
        other => Err(mlua::Error::external(format!(
            "mcp error must be a string or table, got {}",
            other.type_name()
        ))),
    }
}

/// Install `pmacs.mcp.*` (T M9.1).
///
/// Function inventory (10 entries; each parallels the same-named
/// `pmacs.lsp.*` entry, with the single addition of `capabilities`
/// for M9.1's "discoverable through the worker" acceptance bullet):
///
/// | `pmacs.mcp.*`        | `pmacs.lsp.*`        | Notes                                   |
/// |----------------------|----------------------|-----------------------------------------|
/// | `spawn`              | `spawn`              | Same shape; MCP spec drops `language_id`/`root_uri`. |
/// | `stop`               | `stop`               | Identical.                              |
/// | `send_request`       | `send_request`       | Identical (JSON-RPC 2.0 layer is same). |
/// | `send_notification`  | `send_notification`  | Identical.                              |
/// | `send_response`      | `send_response`      | Identical.                              |
/// | `events_take`        | `events_take`        | Identical.                              |
/// | `list`               | `list`               | Identical.                              |
/// | `forget`             | `forget`             | Identical.                              |
/// | `_tick`              | `_tick`              | Identical.                              |
/// | `capabilities`       | (none — folded into `list`/`status_summary`) | M9.1 acceptance: `pmacs.mcp.capabilities(id)` returns the server's declared capabilities table or nil. |
#[allow(
    clippy::too_many_lines,
    reason = "linear list of raw bindings; splitting fragments a coherent surface"
)]
pub fn install_mcp(lua: &Lua, manager: &SharedMcpManager) -> mlua::Result<()> {
    lua.set_app_data(manager.clone());
    let pmacs: Table = lua.globals().get("pmacs")?;
    let mcp_mod = lua.create_table()?;

    {
        let m = manager.clone();
        mcp_mod.set(
            "spawn",
            lua.create_function(move |_, spec: Table| {
                let parsed = lua_to_mcp_spec(&spec)?;
                let id = m
                    .borrow_mut()
                    .spawn(parsed)
                    .map_err(mlua::Error::external)?;
                Ok(McpServerIdLua(id))
            })?,
        )?;
    }

    {
        let m = manager.clone();
        mcp_mod.set(
            "stop",
            lua.create_function(move |_, id: McpServerIdLua| {
                m.borrow_mut().stop(id.0).map_err(mlua::Error::external)?;
                Ok(())
            })?,
        )?;
    }

    {
        // Raw MCP request dispatcher. Returns the async-runtime
        // [`JobId`] the response will settle. Package code does not
        // call this directly — `builtin/runtime/mcp.lua` wraps it
        // into a Handle-returning `pmacs.mcp.send_request` that
        // matches the dispatch shape of `pmacs.workers.compute_sum`,
        // `pmacs.fs.read_dir`, etc. The underscore-prefixed name
        // marks this as the unwrapped primitive (mirrors
        // `pmacs._async._dispatch_*` in async.lua).
        let m = manager.clone();
        mcp_mod.set(
            "_send_request_raw",
            lua.create_function(
                move |_, (id, method, params): (McpServerIdLua, String, Option<Value>)| {
                    let json_params = match params {
                        Some(Value::Nil) | None => serde_json::Value::Null,
                        Some(other) => lua_to_json(other)?,
                    };
                    let job_id = m
                        .borrow_mut()
                        .send_request(id.0, method, json_params)
                        .map_err(mlua::Error::external)?;
                    Ok(job_id)
                },
            )?,
        )?;
    }

    {
        let m = manager.clone();
        mcp_mod.set(
            "send_notification",
            lua.create_function(
                move |_, (id, method, params): (McpServerIdLua, String, Option<Value>)| {
                    let json_params = match params {
                        Some(Value::Nil) | None => serde_json::Value::Null,
                        Some(other) => lua_to_json(other)?,
                    };
                    m.borrow_mut()
                        .send_notification(id.0, method, json_params)
                        .map_err(mlua::Error::external)?;
                    Ok(())
                },
            )?,
        )?;
    }

    {
        let m = manager.clone();
        mcp_mod.set(
            "send_response",
            lua.create_function(
                move |_,
                      (id, request_id, result, err): (
                    McpServerIdLua,
                    Value,
                    Value,
                    Option<Value>,
                )| {
                    let request_id_json = lua_to_json(request_id)?;
                    let outcome = match err {
                        Some(Value::Nil) | None => Ok(lua_to_json(result)?),
                        Some(e) => Err(lua_to_mcp_error(e)?),
                    };
                    m.borrow_mut()
                        .send_response(id.0, request_id_json, outcome)
                        .map_err(mlua::Error::external)?;
                    Ok(())
                },
            )?,
        )?;
    }

    {
        let m = manager.clone();
        mcp_mod.set(
            "events_take",
            lua.create_function(move |lua, id: McpServerIdLua| {
                let evs = m.borrow_mut().take_events(id.0);
                let out = lua.create_table_with_capacity(evs.len(), 0)?;
                for (i, ev) in evs.iter().enumerate() {
                    out.set(i + 1, mcp_event_to_lua(lua, ev)?)?;
                }
                Ok(out)
            })?,
        )?;
    }

    {
        let m = manager.clone();
        mcp_mod.set(
            "list",
            lua.create_function(move |lua, ()| {
                let mgr = m.borrow();
                let ids: Vec<McpServerId> = mgr.ids().collect();
                let out = lua.create_table_with_capacity(ids.len(), 0)?;
                for (i, id) in ids.iter().enumerate() {
                    let row = lua.create_table_with_capacity(0, 4)?;
                    row.set("id", McpServerIdLua(*id))?;
                    if let Some(spec) = mgr.spec(*id) {
                        row.set("label", spec.label.as_str())?;
                        row.set("command", spec.command.as_str())?;
                    }
                    if let Some(state) = mgr.state(*id) {
                        row.set("state", mcp_state_to_lua(lua, state)?)?;
                    }
                    if let Some(attempt) = mgr.attempt(*id) {
                        row.set("attempt", attempt)?;
                    }
                    out.set(i + 1, row)?;
                }
                Ok(out)
            })?,
        )?;
    }

    {
        let m = manager.clone();
        mcp_mod.set(
            "forget",
            lua.create_function(move |_, id: McpServerIdLua| {
                m.borrow_mut().forget(id.0).map_err(mlua::Error::external)?;
                Ok(())
            })?,
        )?;
    }

    {
        let m = manager.clone();
        mcp_mod.set(
            "_tick",
            lua.create_function(move |_, ()| {
                m.borrow_mut().tick();
                Ok(())
            })?,
        )?;
    }

    {
        // M9.1 acceptance: declared capabilities are discoverable
        // through the worker. Returns the server's `capabilities`
        // table (as JSON-translated Lua) or `nil` if the server is
        // not yet initialized.
        let m = manager.clone();
        mcp_mod.set(
            "capabilities",
            lua.create_function(move |lua, id: McpServerIdLua| {
                let mgr_ref = m.borrow();
                match mgr_ref.capabilities(id.0) {
                    Some(caps) => json_to_lua(lua, caps),
                    None => Ok(Value::Nil),
                }
            })?,
        )?;
    }

    {
        // T M9.2: raw resource-read dispatcher. Returns the
        // async-runtime [`JobId`] the response will settle.
        // `builtin/runtime/mcp.lua` wraps it into a Handle-returning
        // `pmacs.mcp.read_resource(server, uri)`.
        let m = manager.clone();
        mcp_mod.set(
            "_read_resource_raw",
            lua.create_function(move |_, (id, uri): (McpServerIdLua, String)| {
                let job_id = m
                    .borrow_mut()
                    .read_resource(id.0, uri)
                    .map_err(mlua::Error::external)?;
                Ok(job_id)
            })?,
        )?;
    }

    {
        // T M9.2: explicit cache invalidation. Drops the per-(server,
        // uri) cache entry; subsequent `read_resource` calls
        // re-dispatch. In-flight requests at the moment of
        // invalidation still settle their awaiters with the arriving
        // result, but do not re-cache.
        let m = manager.clone();
        mcp_mod.set(
            "invalidate_resource",
            lua.create_function(move |_, (id, uri): (McpServerIdLua, String)| {
                m.borrow_mut().invalidate_resource(id.0, uri);
                Ok(())
            })?,
        )?;
    }

    {
        // T M9.3: raw tool dispatcher. Returns the async-runtime
        // [`JobId`] the response will settle. `builtin/runtime/mcp.lua`
        // wraps it into a Handle-returning
        // `pmacs.mcp.invoke_tool(server, name, args)`.
        let m = manager.clone();
        mcp_mod.set(
            "_invoke_tool_raw",
            lua.create_function(
                move |_, (id, name, args): (McpServerIdLua, String, Option<Value>)| {
                    let json_args = match args {
                        Some(Value::Nil) | None => {
                            serde_json::Value::Object(serde_json::Map::default())
                        }
                        Some(other) => lua_to_json(other)?,
                    };
                    let job_id = m
                        .borrow_mut()
                        .invoke_tool(id.0, name, json_args)
                        .map_err(mlua::Error::external)?;
                    Ok(job_id)
                },
            )?,
        )?;
    }

    {
        // T M9.5: register interest in a JSON-RPC notification method.
        // The notification dispatcher in `builtin/runtime/mcp.lua`
        // calls this when the first handler for a method is added,
        // and `_unsubscribe_notification` when the last one drops.
        let m = manager.clone();
        mcp_mod.set(
            "_subscribe_notification",
            lua.create_function(move |_, method: String| {
                m.borrow_mut().subscribe_notification(method);
                Ok(())
            })?,
        )?;
    }

    {
        let m = manager.clone();
        mcp_mod.set(
            "_unsubscribe_notification",
            lua.create_function(move |_, method: String| {
                m.borrow_mut().unsubscribe_notification(&method);
                Ok(())
            })?,
        )?;
    }

    {
        // T M9.5: drain all queued subscribed notifications. Returns
        // a Lua table { [method] = { { server = ..., params = ... }, ... } }.
        // The Lua tick hook walks this and invokes registered
        // handlers.
        let m = manager.clone();
        mcp_mod.set(
            "_drain_notifications",
            lua.create_function(move |lua, ()| {
                let drained = m.borrow_mut().drain_notifications();
                let out = lua.create_table_with_capacity(0, drained.len())?;
                for (method, entries) in drained {
                    let arr = lua.create_table_with_capacity(entries.len(), 0)?;
                    for (i, (sid, params)) in entries.into_iter().enumerate() {
                        let entry = lua.create_table_with_capacity(0, 2)?;
                        entry.set("server", McpServerIdLua(sid))?;
                        entry.set("params", json_to_lua(lua, &params)?)?;
                        arr.set(i + 1, entry)?;
                    }
                    out.set(method.as_str(), arr)?;
                }
                Ok(out)
            })?,
        )?;
    }

    {
        // T M9.4: raw prompt dispatcher. Returns the async-runtime
        // [`JobId`] the response will settle. `builtin/runtime/mcp.lua`
        // wraps it into a Handle-returning
        // `pmacs.mcp.get_prompt(server, name, args)`.
        //
        // `args` of `nil` (or omitted) translates to an empty object
        // on the wire — the MCP spec requires the `arguments` field
        // even for prompts that take no arguments. Three Lua call
        // patterns produce identical wire requests:
        //   pmacs.mcp.get_prompt(s, "p")        (no third arg)
        //   pmacs.mcp.get_prompt(s, "p", nil)   (explicit nil)
        //   pmacs.mcp.get_prompt(s, "p", {})    (empty table)
        let m = manager.clone();
        mcp_mod.set(
            "_get_prompt_raw",
            lua.create_function(
                move |_, (id, name, args): (McpServerIdLua, String, Option<Value>)| {
                    let json_args = match args {
                        Some(Value::Nil) | None => {
                            serde_json::Value::Object(serde_json::Map::default())
                        }
                        Some(other) => lua_to_json(other)?,
                    };
                    let job_id = m
                        .borrow_mut()
                        .get_prompt(id.0, name, json_args)
                        .map_err(mlua::Error::external)?;
                    Ok(job_id)
                },
            )?,
        )?;
    }

    pmacs.set("mcp", mcp_mod)?;
    Ok(())
}

/// Build a fresh [`McpManager`] over `supervisor` and install
/// `pmacs.mcp.*` over it. Mirrors [`make_lsp_manager`] in shape;
/// also takes the editor's [`SharedAsyncRuntime`] so MCP responses
/// settle Lua-visible job ids without occupying a worker thread
/// (T M9.1, Pass-2 finding 1).
pub fn make_mcp_manager(
    lua: &Lua,
    supervisor: SharedProcessSupervisor,
    runtime: crate::async_runtime::SharedAsyncRuntime,
) -> mlua::Result<SharedMcpManager> {
    let manager = Rc::new(RefCell::new(McpManager::new(supervisor, runtime)));
    install_mcp(lua, &manager)?;
    Ok(manager)
}
